File: src/main/java/at/searles/parsing/combinators/RecognizerRep.java
package at.searles.parsing.combinators;

import at.searles.parsing.PrinterCallBack;
import at.searles.parsing.Recognizable;
import at.searles.parsing.Recognizer;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

/**
 * Possibly empty repetition
 */
public class RecognizerRep implements Recognizer, Recognizable.Rep {

    private final Recognizer parent;

    public RecognizerRep(Recognizer parent) {
        this.parent = parent;
    }

    @Override
    public Recognizable parent() {
        return parent;
    }

    @NotNull
    @Override
    public ConcreteSyntaxTree print(PrinterCallBack env) {
        return ConcreteSyntaxTree.empty();
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/RecognizerRep.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.printing.EmptyTree

/**
 * Possibly empty repetition
 */
class RecognizerRep(private val parent: Recognizer) : Recognizer {
    override fun recognize(stream: ParserStream): Boolean {
        if(stream.recognize(parent, true)) {
            while(stream.recognize(parent, false)) {
                // empty
            }
        }

        return true
    }

    override fun print(): ConcreteSyntaxTree {
        return EmptyTree
    }

    override fun toString(): String {
        return "$parent.rep"
    }
}